File: ScaleformUI_Lua/src/Menus/PauseMenus/LobbyMenu/Items/MinimapRoute.lua
MinimapRoute = setmetatable({}, MinimapRoute)
MinimapRoute.__index = MinimapRoute
MinimapRoute.__call = function()
    return "LobbyComponent", "MinimapRoute"
end

---@class MinimapRoute
---@field private New fun()
---@field public StartPoint MinimapRaceCheckpoint
---@field public EndPoint MinimapRaceCheckpoint
---@field public CheckPoints MinimapRaceCheckpoint[]
---@field public RadarThickness number
---@field public MapThickness number
---@field public RouteColor HudColours
---@field private SetupCustomRoute fun()

function MinimapRoute.New()
    local _data = {
        StartPoint = MinimapRaceCheckpoint.New(0, vector3(0, 0, 0)),
        EndPoint = MinimapRaceCheckpoint.New(0, vector3(0, 0, 0)),
        CheckPoints = {},
        MapThickness = 30,
        RouteColor = HudColours.HUD_COLOUR_FREEMODE
    }
    return setmetatable(_data, MinimapRoute)
end

function MinimapRoute:SetupCustomRoute()
    if self.StartPoint == nil or self.StartPoint.Position.x == 0 or self.StartPoint.Position.y == 0 or self.StartPoint.Position.z == 0 then
        return
    end

    ClearGpsFlags()
    SetGpsFlags(8, 0.0)
    StartGpsCustomRoute(self.RouteColor, true, true)

    RaceGalleryNextBlipSprite(self.StartPoint.Sprite)
    RaceGalleryAddBlip(self.StartPoint.Position.x, self.StartPoint.Position.y, self.StartPoint.Position.z)

    AddPointToGpsCustomRoute(self.StartPoint.Position.x, self.StartPoint.Position.y, self.StartPoint.Position.z)

    for i = 1, #self.CheckPoints, 1 do
        local checkPoint = self.CheckPoints[i]
        RaceGalleryNextBlipSprite(checkPoint.Sprite)
        local b = RaceGalleryAddBlip(checkPoint.Position.x, checkPoint.Position.y, checkPoint.Position.z)
        if checkPoint.Scale > 0 then
            SetBlipScale(b, checkPoint.Scale)
        end
        SetBlipColour(b, checkPoint.Color)
        AddPointToGpsCustomRoute(checkPoint.Position.x, checkPoint.Position.y, checkPoint.Position.z)
    end

    RaceGalleryNextBlipSprite(self.EndPoint.Sprite)
    RaceGalleryAddBlip(self.EndPoint.Position.x, self.EndPoint.Position.y, self.EndPoint.Position.z)
    AddPointToGpsCustomRoute(self.EndPoint.Position.x, self.EndPoint.Position.y, self.EndPoint.Position.z)

    SetGpsCustomRouteRender(true, 18, self.MapThickness)
end

File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Items/MinimapRoute.cs
using ScaleformUI.Scaleforms;
using static CitizenFX.Core.Native.API;

namespace ScaleformUI.PauseMenus.Elements.Items
{
    [Flags]
    public enum GPSFlags
    {
        NONE = 0,
        IGNORE_ONE_WAY = 1,
        FOLLOW_RULES = 2,
        AVOID_HIGHWAY = 4,
        NO_ROUTE_SHIFT = 8,
        CUSTOM_PROXIMITY = 16,
        NO_PULL_PATH_TO_RIGHT_LANE = 32,
        AVOID_OFF_ROAD = 64,
        IGNORE_DESTINATION_Z = 128
    }

    public class MinimapRoute : MinimapBaseItem
    {
        public MinimapRaceCheckpoint StartPoint = new();
        public MinimapRaceCheckpoint EndPoint = new();
        public List<MinimapRaceCheckpoint> CheckPoints = new List<MinimapRaceCheckpoint>();
        public int RadarThickness = 18;
        public int MapThickness = 30;
        public bool FollowStreet = true;
        public HudColor RouteColor { get; set; } = HudColor.HUD_COLOUR_FREEMODE;
        public void SetupCustomRoute()
        {
            if (StartPoint.Position.IsZero) return;
            ClearGpsFlags();
            SetGpsFlags(8, 0f);
            StartGpsCustomRoute((int)RouteColor, true, true);

            RaceGalleryNextBlipSprite(StartPoint.BlipSprite);
            RaceGalleryAddBlip(StartPoint.Position.X, StartPoint.Position.Y, StartPoint.Position.Z);

            AddPointToGpsCustomRoute(StartPoint.Position.X, StartPoint.Position.Y, StartPoint.Position.Z);

            for (int i = 0; i < CheckPoints.Count; i++)
            {
                MinimapRaceCheckpoint checkPoint = CheckPoints[i];
                RaceGalleryNextBlipSprite(checkPoint.BlipSprite);
                int blip = RaceGalleryAddBlip(checkPoint.Position.X, checkPoint.Position.Y, checkPoint.Position.Z);
                if (checkPoint.Scale > 0)
                    SetBlipScale(blip, checkPoint.Scale);
                SetBlipColour(blip, (int)checkPoint.Color);
                ShowNumberOnBlip(blip, i)
                AddPointToGpsCustomRoute(checkPoint.Position.X, checkPoint.Position.Y, checkPoint.Position.Z);
            }

            RaceGalleryNextBlipSprite(EndPoint.BlipSprite);
            RaceGalleryAddBlip(EndPoint.Position.X, EndPoint.Position.Y, EndPoint.Position.Z);
            AddPointToGpsCustomRoute(EndPoint.Position.X, EndPoint.Position.Y, EndPoint.Position.Z);

            SetGpsCustomRouteRender(true, 18, MapThickness); ;
        }
    }
}
